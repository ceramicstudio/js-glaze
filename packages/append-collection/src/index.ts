export * from './append-collection'
export * from './constants'
export * from './cursor'
export * from './schemas'
export * from './types'
