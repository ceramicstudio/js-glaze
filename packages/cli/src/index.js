/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-var-requires */

const { run } = require('@oclif/command')
const updateNotifier = require('update-notifier')

const pkg = require('../package.json')
updateNotifier({ pkg }).notify()

exports.run = run
