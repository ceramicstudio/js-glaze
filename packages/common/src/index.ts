import type { JSONSchemaType } from 'ajv'

export type Definition<C extends Record<string, any> = Record<string, any>> = {
  name: string
  description: string
  schema: string
  url?: string
  config?: C
}

export type Schema<T = Record<string, any>> = JSONSchemaType<T> & {
  $comment?: string
  title?: string
}

export type ModelData<T> = {
  definitions: Record<string, T>
  schemas: Record<string, T>
  tiles: Record<string, T>
}

export type PublishedModel = ModelData<string>

export const CIP88_APPEND_COLLECTION_PREFIX = 'cip88:appendCollection:'
export const CIP88_REF_PREFIX = 'cip88:ref:'
