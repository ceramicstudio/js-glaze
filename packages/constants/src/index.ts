export const CIP11_DEFINITION_SCHEMA_URL =
  'ceramic://k3y52l7qbv1fry1fp4s0nwdarh0vahusarpposgevy0pemiykymd2ord6swtharcw'
export const CIP11_INDEX_SCHEMA_URL =
  'ceramic://k3y52l7qbv1fryjn62sggjh1lpn11c56qfofzmty190d62hwk1cal1c7qc5he54ow'

export const CIP88_APPEND_COLLECTION_PREFIX = 'cip88:appendCollection:'
export const CIP88_REF_PREFIX = 'cip88:ref:'
