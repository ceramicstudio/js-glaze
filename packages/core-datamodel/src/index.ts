import type { ModelTypeAliases, EncodedSignedModel } from '@glazed/types'

export type Definition<C extends Record<string, any> = Record<string, any>> = {
  name: string
  description: string
  schema: string
  url?: string
  config?: C
}

export type IdentityIndex = Record<string, string>

export type ModelTypes = ModelTypeAliases<{
  Definition: Definition
  IdentityIndex: IdentityIndex
}>

export const model: EncodedSignedModel = {
  definitions: {},
  schemas: {
    Definition: [
      {
        jws: {
          payload: 'AXESIGVBGu8EHXq5fxC16J4MwEPBCfrWq1bsZWQP2F_WWhT0',
          signatures: [
            {
              protected:
                'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBI3o2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QSJ9',
              signature:
                'hvv_KWbyYbg7ZP8PQeJxPjHnlAg1VKtPjVa6zXwMq9pug96qGcXFQi5XUAl30RH4QYOHl8Gpv4uZczWxAJllAg',
            },
          ],
          link: 'bafyreidfieno6ba5pk4x6efv5cpazqcdyee7vvvlk3wgkzap3bp5mwqu6q',
        },
        linkedBlock:
          'o2RkYXRhpmR0eXBlZm9iamVjdGV0aXRsZWpEZWZpbml0aW9uZyRzY2hlbWF4J2h0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hI2hyZXF1aXJlZINkbmFtZWtkZXNjcmlwdGlvbmZzY2hlbWFqcHJvcGVydGllc6VjdXJsomR0eXBlZnN0cmluZ2ltYXhMZW5ndGgY8GRuYW1lomR0eXBlZnN0cmluZ2ltYXhMZW5ndGgYlmZjb25maWehZHR5cGVmb2JqZWN0ZnNjaGVtYaFkJHJlZngaIy9kZWZpbml0aW9ucy9DZXJhbWljRG9jSWRrZGVzY3JpcHRpb26iZHR5cGVmc3RyaW5naW1heExlbmd0aBkBpGtkZWZpbml0aW9uc6FsQ2VyYW1pY0RvY0lko2R0eXBlZnN0cmluZ2dwYXR0ZXJueBxeY2VyYW1pYzovLy4rKFw/dmVyc2lvbj0uKyk/aW1heExlbmd0aBiWZmhlYWRlcqJmc2NoZW1h92tjb250cm9sbGVyc4F4OGRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBZ2RvY3R5cGVkdGlsZQ==',
      },
    ],
    IdentityIndex: [
      {
        jws: {
          payload: 'AXESIKO29HNjXsFTO9Bbs5VTTGqJUPZ9Aoic83L6G0ziXp48',
          signatures: [
            {
              protected:
                'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBI3o2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QSJ9',
              signature:
                'R-VH5EwcOsvd3Txsl4Pjw-QSobjiNJuQVPWrbubyX6uIbe2SRupKx5od0z2agdeVnglt9L9mHiOqlRGgI7WNDQ',
            },
          ],
          link: 'bafyreifdw32hgy26yfjtxuc3wokvgtdkrfipm7icrcopg4x2dngoexu6hq',
        },
        linkedBlock:
          'o2RkYXRhpWR0eXBlZm9iamVjdGV0aXRsZW1JZGVudGl0eUluZGV4ZyRzY2hlbWF4J2h0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hI2tkZWZpbml0aW9uc6FsQ2VyYW1pY0RvY0lko2R0eXBlZnN0cmluZ2dwYXR0ZXJueBxeY2VyYW1pYzovLy4rKFw/dmVyc2lvbj0uKyk/aW1heExlbmd0aBiWdGFkZGl0aW9uYWxQcm9wZXJ0aWVzoWQkcmVmeBojL2RlZmluaXRpb25zL0NlcmFtaWNEb2NJZGZoZWFkZXKiZnNjaGVtYfdrY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QWdkb2N0eXBlZHRpbGU=',
      },
    ],
  },
  tiles: {},
}
