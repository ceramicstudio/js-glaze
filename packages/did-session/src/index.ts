/**
 * Manages user account and DID in web based environments.
 *
 * ## Purpose
 *
 * Manages, creates and authorizes a DID session key for a user. Returns an authenticated DIDs instance
 * to be used in other Ceramic libraries. Supports did:pkh for blockchain accounts with Sign-In with
 * Ethereum and CACAO for authorization.
 *
 * ## Installation
 *
 * ```sh
 * npm install @glazed/did-session
 * ```
 *
 * ## Usage
 *
 * Create an instance, authorize and use DIDs where needed. At the moment, only Ethereum accounts
 * are supported with the EthereumAuthProvider. Additional accounts will be supported in the future.
 *
 * ```ts
 * import { DIDSession } from '@glazed/did-session'
 * import { EthereumAuthProvider } from '@ceramicnetwork/blockchain-utils-linking'
 *
 * const ethProvider = // import/get your web3 eth provider
 * const addresses = await ethProvider.enable()
 * const authProvider = new EthereumAuthProvider(ethProvider, addresses[0])
 *
 * const session = new DIDSession({ authProvider })
 * const did = await session.authorize()
 *
 * // Uses DIDs in ceramic & glaze libraries, ie
 * const ceramic = new CeramicClient()
 * ceramic.did = did
 *
 * // pass ceramic instance where needed
 *
 * ```
 *
 * You can serialize a session to store for later and then re-initialize. Currently sessions are valid
 * for 1 day by default.
 *
 *```ts
 * // Create session as above, store for later
 * const session = new DIDSession({ authProvider })
 * const did = await session.authorize()
 * const sessionString = session.serialize()
 *
 * // write/save session string where you want (ie localstorage)
 * // ...
 *
 * // Later re initialize session
 * const session2 = DIDSession.fromSession(authProvider, sessionString)
 * const ceramic = new CeramicClient()
 * ceramic.did = session2.getDID()
 * ```
 *
 * Additional helper functions are available to help you manage a session lifecycle and the user experience.
 *
 *  *```ts
 * // Check if authorized or created from existing session string
 * didsession.hasSession
 *
 * // Check if session expired
 * didsession.isExpired
 *
 * // Get resources session is authorized for
 * didsession.authorizations
 *
 * // Check number of seconds till expiration, may want to re auth user at a time before expiration
 * didsession.expiresInSecs
 *
 * @module did-session
 */

import { Ed25519Provider } from 'key-did-provider-ed25519'
import KeyDidResolver from 'key-did-resolver'
import { randomBytes } from '@stablelib/random'
import { DID } from 'dids'
import type { EthereumAuthProvider, CapabilityOpts } from '@ceramicnetwork/blockchain-utils-linking'
import type { Cacao } from 'ceramic-cacao'
import * as u8a from 'uint8arrays'

export type SessionParams = {
  /**
   * An authProvider for the chain you wish to support, only ETH supported at moment
   */
  authProvider: EthereumAuthProvider
  resources?: Array<string>
  keySeed?: Uint8Array
  cacao?: Cacao
}

type SessionObj = {
  sessionKeySeed: string
  cacao: Cacao
}

export async function createDIDKey(seed?: Uint8Array): Promise<DID> {
  const didProvider = new Ed25519Provider(seed || randomBytes(32))
  const didKey = new DID({
    provider: didProvider,
    resolver: KeyDidResolver.getResolver(),
  })
  await didKey.authenticate()
  return didKey
}

export function JSONToBase64url(object: Record<string, any>): string {
  return u8a.toString(u8a.fromString(JSON.stringify(object)), 'base64url')
}

export function base64urlToJSON(s: string): Record<string, any> {
  return JSON.parse(u8a.toString(u8a.fromString(s, 'base64url'))) as Record<string, any>
}

export function bytesToBase64(b: Uint8Array): string {
  return u8a.toString(b, 'base64pad')
}

export function base64ToBytes(s: string): Uint8Array {
  return u8a.fromString(s, 'base64pad')
}

/**
 * DID Session
 *
 * ```sh
 * import { DIDSession } from '@glazed/did-session'
 * ```
 */
export class DIDSession {
  #authProvider: EthereumAuthProvider
  #resources: Array<string>
  #did?: DID
  #keySeed?: Uint8Array
  #cacao?: Cacao

  constructor(params: SessionParams) {
    this.#authProvider = params.authProvider
    this.#keySeed = params.keySeed
    this.#cacao = params.cacao
    this.#resources = params.resources ?? [`ceramic://*`]
  }

  /**
   * Get authProvider
   */
  get authProvider() {
    return this.#authProvider
  }

  /**
   * Request authorization for session
   */
  async authorize(capabilityOpts: CapabilityOpts = {}): Promise<DID> {
    this.#keySeed = randomBytes(32)
    const didKey = await createDIDKey(this.#keySeed)
    // Pass through opts resources instead, resource arg does not support anything but streamids at moment
    const opts = Object.assign({ resources: this.#resources }, capabilityOpts)
    this.#cacao = await this.#authProvider.requestCapability(didKey.id, [], opts)
    return this.initDID(didKey, this.#cacao)
  }

  async initDID(didKey: DID, cacao: Cacao): Promise<DID> {
    const didWithCap = didKey.withCapability(cacao)
    await didWithCap.authenticate()
    this.#did = didWithCap
    return didWithCap
  }

  /**
   * Get DID instance, if authorized
   */
  getDID(): DID {
    if (!this.#did) {
      throw new Error('DID not available, has not authorized')
    }
    return this.#did
  }

  /**
   * Serialize session into string, can store and initalize the same session again while valid
   */
  serialize(): string {
    if (!this.#keySeed || !this.#cacao) throw new Error('No session to seralize')
    const session = {
      sessionKeySeed: bytesToBase64(this.#keySeed),
      cacao: this.#cacao,
    }
    return JSONToBase64url(session)
  }

  /**
   * Initialize a session from a serialized session string
   */
  static async fromSession(
    session: string,
    authProvider: EthereumAuthProvider
  ): Promise<DIDSession> {
    const { sessionKeySeed, cacao } = base64urlToJSON(session) as SessionObj
    const dsession = new DIDSession({ authProvider, cacao, keySeed: base64ToBytes(sessionKeySeed) })
    const didKey = await createDIDKey(dsession.#keySeed)
    await dsession.initDID(didKey, cacao)
    return dsession
  }

  get hasSession(): boolean {
    return !!this.#cacao && !!this.#did
  }

  /**
   * Determine if a session is expired or not
   */
  get isExpired(): boolean {
    if (!this.#cacao) throw new Error('No session available')
    const expTime = this.#cacao.p.exp
    if (!expTime) return false
    return Date.parse(expTime) < Date.now()
  }

  /**
   * Number of seconds until a session expires
   */
  get expireInSecs(): number {
    if (!this.#cacao) throw new Error('No session available')
    const expTime = this.#cacao.p.exp
    if (!expTime) throw new Error('Session does not expire')
    const timeDiff = Date.parse(expTime) - Date.now()
    return timeDiff < 0 ? 0 : timeDiff / 1000
  }

  /**
   * Get the list of resources a session is authorized for
   */
  get authorizations(): Array<string> {
    return this.#cacao?.p.resources ?? []
  }

  /**
   * Get the session CACAO
   */
  get cacao(): Cacao {
    if (!this.#cacao) throw new Error('No session available')
    return this.#cacao
  }

  /**
   * Determine if session is available and optionally if authorized for given resources
   */
  isAuthorized(resources?: Array<string>): boolean {
    if (!this.hasSession || this.isExpired) return false
    if (!resources) return true
    return resources.every((val) => this.authorizations.includes(val))
  }

  /** DID string associated to the session instance. session.id == session.getDID().parent */
  get id(): string {
    if (!this.#did) {
      throw new Error('ID not available, has not authorized')
    }
    return this.#did.parent
  }
}
