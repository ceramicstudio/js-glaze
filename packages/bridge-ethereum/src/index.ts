/**
 * Ethereum bridge
 *
 * @module bridge-ethereum
 */

import { type CapabilityOpts, EthereumAuthProvider } from '@ceramicnetwork/blockchain-utils-linking'
import type { CeramicApi } from '@ceramicnetwork/common'
import { CeramicClient } from '@ceramicnetwork/http-client'
import { DIDSession } from '@glazed/did-session'
import { GraphQLClient } from '@glazed/graphql'
import type { RuntimeCompositeDefinition } from '@glazed/types'
import type { ExecutionResult, Source } from 'graphql'

export type EthereumBridgeParams = {
  authProvider: EthereumAuthProvider
  ceramic: CeramicApi | string
  definition: RuntimeCompositeDefinition
}

export class EthereumBridge {
  #client: GraphQLClient
  #resources: Array<string>
  #session: DIDSession

  // TODO: fromEthereumProvider() using an EthereumProvider and creating the authProvider

  constructor(params: EthereumBridgeParams) {
    const ceramic =
      typeof params.ceramic === 'string' ? new CeramicClient(params.ceramic) : params.ceramic
    this.#client = GraphQLClient.fromDefinition({ ceramic, definition: params.definition })
    this.#resources = Object.values(params.definition.models).map((modelID) => {
      // TODO: is this the right syntax?
      return `ceramic://*?model=${modelID}`
    })
    this.#session = new DIDSession({ authProvider: params.authProvider })
  }

  get id(): string {
    return this.#session.id
  }

  async authorize(options: CapabilityOpts = {}): Promise<string> {
    const did = await this.#session.authorize({ resources: this.#resources, ...options })
    this.#client.context.ceramic.did = did
    return did.id
  }

  async execute(
    source: string | Source,
    variables?: Record<string, unknown>
  ): Promise<ExecutionResult> {
    return await this.#client.execute(source, variables)
  }
}
