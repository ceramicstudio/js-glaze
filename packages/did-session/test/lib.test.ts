/**
 * @jest-environment glaze
 */
import type { CeramicApi } from '@ceramicnetwork/common'
import { TileDocument } from '@ceramicnetwork/stream-tile'
import { EventEmitter } from 'events'
import { EthereumAuthProvider } from '@ceramicnetwork/blockchain-utils-linking'
import { Wallet as EthereumWallet } from '@ethersproject/wallet'
import { fromString, toString } from 'uint8arrays'
import { DIDSession } from '../src'
import { jest } from '@jest/globals'
import { Wallet } from '@ethersproject/wallet'
import { SiweMessage, Cacao } from 'ceramic-cacao'
import { Model, ModelAccountRelation, ModelDefinition } from '@ceramicnetwork/stream-model'
import { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'

const getModelDef = (name: string): ModelDefinition => ({
  name: name,
  accountRelation: ModelAccountRelation.LIST,
  schema: {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    type: 'object',
    additionalProperties: false,
    properties: {
      myData: {
        type: 'integer',
        maximum: 10000,
        minimum: 0,
      },
    },
    required: ['myData'],
  },
})

const CONTENT0 = { myData: 0 }
const CONTENT1 = { myData: 1 }

const MODEL_DEFINITION = getModelDef('MyModel')

class EthereumProvider extends EventEmitter {
  wallet: EthereumWallet

  constructor(wallet: EthereumWallet) {
    super()
    this.wallet = wallet
  }

  send(
    request: { method: string; params: Array<any> },
    callback: (err: Error | null | undefined, res?: any) => void
  ): void {
    if (request.method === 'eth_chainId') {
      callback(null, { result: '1' })
    } else if (request.method === 'personal_sign') {
      let message = request.params[0] as string
      if (message.startsWith('0x')) {
        message = toString(fromString(message.slice(2), 'base16'), 'utf8')
      }
      callback(null, { result: this.wallet.signMessage(message) })
    } else {
      callback(new Error(`Unsupported method: ${request.method}`))
    }
  }
}

function createEthereumAuthProvider(mnemonic?: string): Promise<EthereumAuthProvider> {
  const wallet = mnemonic ? EthereumWallet.fromMnemonic(mnemonic) : EthereumWallet.createRandom()
  const provider = new EthereumProvider(wallet)
  return Promise.resolve(new EthereumAuthProvider(provider, wallet.address))
}

const bytes32 = [
  64, 168, 135, 95, 204, 113, 52, 90, 66, 192, 219, 241, 34, 128, 184, 176, 36, 249, 191, 223, 108,
  240, 6, 119, 226, 7, 81, 210, 31, 128, 182, 139,
]

const testResources = [
  '`ceramic://*?model=k2t6wyfsu4pfz0fnidk6tz3gak7tr8n5w34ah1c31w5vq98b62hir1pnn3j2ty',
  '`ceramic://*?model=k2t6wyfsu4pgz0ftx664veuaf2qib95zj8je2x7pf89v6g5p7xa7n9eo45g64a',
]

declare global {
  const ceramic: CeramicApi
}

describe('did-session', () => {
  let authProvider: EthereumAuthProvider
  jest.setTimeout(20000)
  const opts = { domain: 'myApp' }
  let model: Model

  beforeAll(async () => {
    authProvider = await createEthereumAuthProvider()
    model = await Model.create(ceramic, MODEL_DEFINITION)
  })

  const wallet = Wallet.fromMnemonic(
    'despair voyage estate pizza main slice acquire mesh polar short desk lyrics'
  )
  const address = wallet.address

  test('creates did-session', () => {
    const session = new DIDSession({ authProvider })
    expect(session.authProvider).toBe(authProvider)
  })

  test('did getter throws error when not authorized', () => {
    const session = new DIDSession({ authProvider })
    expect(() => session.getDID()).toThrow('DID not available, has not authorized')
  })

  test('did getter returns when authorized', async () => {
    const session = new DIDSession({ authProvider })
    const did = await session.authorize(opts)
    expect(did).toBeTruthy()
    expect(session.getDID()).toBeTruthy()
  })

  test('authorize, default wildcard', async () => {
    const session = new DIDSession({ authProvider })
    const did = await session.authorize(opts)
    expect(did.capability.p.resources.includes(`ceramic://*`)).toBe(true)
  })

  test('authorize, with streamid resources', async () => {
    const session = new DIDSession({ authProvider })
    const streamId = `ceramic://z6MkhZCWzHtPFmpNupVPuHA6svtpKKY9RUpgf9uohnhFMNvj`
    const did = await session.authorize({ resources: [streamId], domain: 'myApp' })
    expect(did.capability.p.resources.includes(streamId)).toBe(true)
  })

  test('authorize and create/update streams', async () => {
    const session = new DIDSession({ authProvider })
    const did = await session.authorize(opts)
    ceramic.did = did
    const doc = await TileDocument.create(
      ceramic,
      { foo: 'bar' },
      {},
      {
        anchor: false,
        publish: false,
      }
    )
    expect(doc.content).toEqual({ foo: 'bar' })

    await doc.update({ foo: 'boo' })
    expect(doc.content).toEqual({ foo: 'boo' })
  })

  // Enable with next release
  test.skip('can create and update model instance stream', async () => {
    const session = new DIDSession({ authProvider, resources: [model.id.toString()] })
    const did = await session.authorize(opts)
    ceramic.did = did

    const doc = await ModelInstanceDocument.create(ceramic, CONTENT0, {
      model: model.id,
    })

    expect(doc.content).toEqual(CONTENT0)
    expect(doc.metadata.model.toString()).toEqual(model.id.toString())

    await doc.replace(CONTENT1)
    expect(doc.content).toEqual(CONTENT1)
  })

  test('isAuthorized/isExpired, with valid session and resources', async () => {
    const session = new DIDSession({ authProvider, resources: testResources })
    await session.authorize({ domain: 'myApp' })
    // Any session authorized and valid, true
    expect(session.isAuthorized()).toBe(true)
    expect(session.isExpired).toBe(false)
    // Authorized for given resources, true
    expect(session.isAuthorized(testResources)).toBe(true)
    // Authorized for wildcard resource, false
    expect(session.isAuthorized([`ceramic://*`])).toBe(false)
  })

  test('isAuthorized/isExpired, with expired session', async () => {
    // Expired 5 min ago
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      expirationTime: new Date(Date.now() - 1000 * 60 * 5).toISOString(),
      chainId: '1',
      resources: testResources,
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature
    const cacao = Cacao.fromSiweMessage(msg)

    const session = new DIDSession({ authProvider, cacao, keySeed: new Uint8Array(bytes32) })
    expect(session.isExpired).toBe(true)
    expect(session.isAuthorized()).toBe(false)
  })

  test('expiresInSecs, when session valid', async () => {
    // Expires in 5 mins
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      expirationTime: new Date(Date.now() + 1000 * 60 * 5).toISOString(),
      chainId: '1',
      resources: testResources,
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature
    const cacao = Cacao.fromSiweMessage(msg)

    const session = new DIDSession({ authProvider, cacao, keySeed: new Uint8Array(bytes32) })

    // 5 sec buffer
    expect(session.expireInSecs).toBeGreaterThan(60 * 5 - 5)
    expect(session.expireInSecs).toBeLessThan(60 * 5 + 5)
  })

  test('expiresInSecs, when session expired', async () => {
    // Expired 5 min ago
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      expirationTime: new Date(Date.now() - 1000 * 60 * 5).toISOString(),
      chainId: '1',
      resources: testResources,
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature
    const cacao = Cacao.fromSiweMessage(msg)

    const session = new DIDSession({ authProvider, cacao, keySeed: new Uint8Array(bytes32) })

    expect(session.expireInSecs).toEqual(0)
  })

  describe('Manage session state', () => {
    test('serializes', async () => {
      const msg = new SiweMessage({
        domain: 'service.org',
        address: address,
        statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
        uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
        version: '1',
        nonce: '32891757',
        issuedAt: '2021-09-30T16:25:24.000Z',
        chainId: '1',
        resources: testResources,
      })

      const signature = await wallet.signMessage(msg.toMessage())
      msg.signature = signature
      const cacao = Cacao.fromSiweMessage(msg)

      const session = new DIDSession({ authProvider, cacao, keySeed: new Uint8Array(bytes32) })
      const sessionStr = session.serialize()
      expect(sessionStr).toMatchSnapshot()
    })

    test('roundtrip serialization, fromSession', async () => {
      const session = new DIDSession({ authProvider })
      await session.authorize(opts)
      const sessionStr = session.serialize()
      const session2 = DIDSession.fromSession(sessionStr, authProvider)
      const sessionStr2 = session2.serialize()
      expect(sessionStr).toEqual(sessionStr2)
    })
  })
})
