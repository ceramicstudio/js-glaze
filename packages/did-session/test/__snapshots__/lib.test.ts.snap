// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`did-session Manage session state serializes 1`] = `"eyJzZXNzaW9uS2V5U2VlZCI6IlFLaUhYOHh4TkZwQ3dOdnhJb0M0c0NUNXY5OXM4QVozNGdkUjBoK0F0b3M9IiwiY2FjYW8iOnsiaCI6eyJ0IjoiZWlwNDM2MSJ9LCJwIjp7ImRvbWFpbiI6InNlcnZpY2Uub3JnIiwiaWF0IjoiMjAyMS0wOS0zMFQxNjoyNToyNC4wMDBaIiwiaXNzIjoiZGlkOnBraDplaXAxNTU6MToweEJkOUQ5YzdEQzM4OTcxNWE4OWZDODE0OUU0YTVCZTkxMzM2QjI3OTYiLCJhdWQiOiJkaWQ6a2V5Ono2TWtyQmROZHdVUG5YRFZEMURDeGVkelZWQnBhR2k4YVNtb1hGQWVLTmd0QWVyOCIsInZlcnNpb24iOiIxIiwibm9uY2UiOiIzMjg5MTc1NyIsInN0YXRlbWVudCI6IkkgYWNjZXB0IHRoZSBTZXJ2aWNlT3JnIFRlcm1zIG9mIFNlcnZpY2U6IGh0dHBzOi8vc2VydmljZS5vcmcvdG9zIiwicmVzb3VyY2VzIjpbImBjZXJhbWljOi8vKj9tb2RlbD1rMnQ2d3lmc3U0cGZ6MGZuaWRrNnR6M2dhazd0cjhuNXczNGFoMWMzMXc1dnE5OGI2MmhpcjFwbm4zajJ0eSIsImBjZXJhbWljOi8vKj9tb2RlbD1rMnQ2d3lmc3U0cGd6MGZ0eDY2NHZldWFmMnFpYjk1emo4amUyeDdwZjg5djZnNXA3eGE3bjllbzQ1ZzY0YSJdfSwicyI6eyJ0IjoiZWlwMTkxIiwicyI6IjB4MjVmNDBjZDhjNTUwZTYyYjBkNzZjMWRmOTA5MGI0YTcwODk4YmQ1MWE2Y2UyYjA2YWJmNzNkNzRjYTRkYTBlZTE0Y2VlYWFhMzU3YWU0NjkzNzEwM2ExNzYzZjAwNDVmZTBkYTc4NTFjZTdiZWU2MTI1ZGNhNjMxYzQ1Y2ZlMTcxYyJ9fX0"`;
